File: src/domain/repositories/ILearnerRepository.ts
import { LearnerProfile } from '../entities/LearnerProfile';
import { SkillScore } from '../entities/SkillScore';

export interface AnswerAttemptRecord {
  id: string;
  userId?: string;
  courseId: string;
  skillId: string;
  lessonId?: string;
  questionId: string;
  selectedIndex: number;
  selectedAnswer: string;
  isCorrect: boolean;
  timestamp: number;
}

export interface PlacementAttemptSubmission {
  questionId: string;
  selectedIndex: number;
  isCorrect: boolean;
}

export interface PlacementAttemptData {
  courseId: string;
  score: number;
  weakestSkillId?: string;
  submissions: PlacementAttemptSubmission[];
}

export interface ILearnerRepository {
  getProfile(): Promise<LearnerProfile>;
  updateProfile(updates: Partial<LearnerProfile>): Promise<LearnerProfile>;
  saveSkillScores(courseId: string, scores: Record<string, SkillScore>): Promise<void>;
  savePlacementAttempt(data: PlacementAttemptData): Promise<string>;
  markLessonCompleted(lessonId: string): Promise<void>;
  markNodeCompleted(nodeId: string): Promise<void>;
  markReinforcementCompleted(reinforcementId: string): Promise<void>;
  recordAnswerAttempt(attempt: Omit<AnswerAttemptRecord, 'id' | 'timestamp'>): Promise<AnswerAttemptRecord>;
  getAnswerAttempts(limit?: number): Promise<AnswerAttemptRecord[]>;
  addXp(amount: number, actionIdempotencyKey?: string): Promise<number>;
  recordDailyActivity(dateStr?: string): Promise<number>;
  resetAll(): Promise<LearnerProfile>;
}

File: mobile/lib/domain/repositories/i_learner_repository.dart
import '../entities/learner_profile.dart';
import '../entities/skill_score.dart';

class AnswerAttemptRecord {
  final String id;
  final String courseId;
  final String skillId;
  final String lessonId;
  final String questionId;
  final int selectedIndex;
  final String selectedAnswer;
  final bool isCorrect;
  final int timestamp;

  const AnswerAttemptRecord({
    required this.id,
    required this.courseId,
    required this.skillId,
    required this.lessonId,
    required this.questionId,
    required this.selectedIndex,
    required this.selectedAnswer,
    required this.isCorrect,
    required this.timestamp,
  });
}

abstract class ILearnerRepository {
  Future<LearnerProfile> getProfile();
  Future<LearnerProfile> updateProfile(LearnerProfile updated);
  Future<void> saveSkillScores(String courseId, Map<String, SkillScore> scores);
  Future<void> markLessonCompleted(String lessonId);
  Future<void> markNodeCompleted(String nodeId);
  Future<void> markReinforcementCompleted(String reinforcementId);
  Future<AnswerAttemptRecord> recordAnswerAttempt({
    required String courseId,
    required String skillId,
    required String lessonId,
    required String questionId,
    required int selectedIndex,
    required String selectedAnswer,
    required bool isCorrect,
  });
  Future<List<AnswerAttemptRecord>> getAnswerAttempts([int limit = 10]);
  Future<int> addXp(int amount, [String? actionIdempotencyKey]);
  Future<int> recordDailyActivity([String? dateStr]);
  Future<LearnerProfile> resetAll();
}
